shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java

```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     restartAlwaysRequired = isRestartAlwaysRequired();
   }
 
+  // Loosely based off of
+  // https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt
+  // When the IME is shown or hidden, it sends an onApplyWindowInsets call with the
+  // final state of the IME. This defers the final call to allow the animation to
+  // take place before re-calling onApplyWindowInsets after animation completion.
+  @VisibleForTesting
+  @TargetApi(30)
+  @RequiresApi(30)
+  @SuppressLint("NewApi")
+  class ImeSyncDeferringInsetsCallback extends WindowInsetsAnimation.Callback
+      implements View.OnApplyWindowInsetsListener {
+    private int overlayInsetTypes;
+    private int deferredInsetTypes;
+
+    private View view;
+    private WindowInsets lastWindowInsets;
+    private boolean deferredInsets = false;
+    private boolean started = false;
+
+    ImeSyncDeferringInsetsCallback(View view, int persistentInsetTypes, int deferredInsetTypes) {
+      super(WindowInsetsAnimation.Callback.DISPATCH_MODE_CONTINUE_ON_SUBTREE);
+      this.overlayInsetTypes = overlayInsetTypes;
+      this.deferredInsetTypes = deferredInsetTypes;
+      this.view = view;
+    }
+
+    @Override
+    public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
+      this.view = view;
+      if (!started) {
```

Comment:
ya, these apis are pretty confusing. I think this needs a lot more comments. Comments at the class level for what the flow should be for the persistent vs delayed stuff are. When this !started should happen, when should deferredInsets happen.

i.e. I don't think the next maintainer would know what to do here if there was a bug and you're out.

Reply:
I'll make a pass and document the whole thing